src/machine/asm/mod.rs

```diff
@@ -101,6 +102,97 @@ fn check_memory(machine: &mut AsmCoreMachine, page_indices: &(u64, u64)) {
     }
 }
 
+// check whether a memory address is writable or not and mark it as dirty, `size` should be 1, 2, 4 or 8
+fn check_memory_writable(
+    machine: &mut Box<AsmCoreMachine>,
+    addr: u64,
+    size: usize,
+) -> Result<(), Error> {
+    debug_assert!(size == 1 || size == 2 || size == 4 || size == 8);
+    let page = addr >> RISCV_PAGE_SHIFTS;
+    if page as usize >= RISCV_PAGESIZE {
+        return Err(Error::OutOfBound);
+    }
+
+    let page_offset = addr as usize % RISCV_PAGESIZE;
+    if page_offset + size <= RISCV_PAGESIZE {
+        let page_flag = machine.fetch_flag(page)?;
+        if (page_flag & FLAG_WXORX_BIT) != (FLAG_WRITABLE & FLAG_WXORX_BIT) {
+            return Err(Error::InvalidPermission);
+        }
+
+        let frame = page >> MEMORY_FRAME_PAGE_SHIFTS;
+        if machine.frames[frame as usize] == 0 {
+            inited_memory(frame, machine);
+            machine.frames[frame as usize] = 1;
+        }
+
+        machine.set_flag(page, FLAG_DIRTY)?
+    } else {
```

Comment:
If a `write` method use two pages, then two pages should be checked, but you only seem to check the latter page here.

Reply:
sorry for the mistake, please review again.